astropy/cosmology/core.py

```diff
@@ -375,11 +388,70 @@ class FlatCosmologyMixin(metaclass=abc.ABCMeta):
     but ``FlatLambdaCDM`` **will** be flat.
     """
 
+    def __init_subclass__(cls: Type[_FlatCT]) -> None:
+        super().__init_subclass__()
+
+        # Determine the non-flat class.
+        # This will raise a TypeError if the MRO is inconsistent.
+        cls._nonflat_cls_
+
+    # ===============================================================
+
+    @classmethod  # TODO! make metaclass-method
+    def _get_nonflat_cls(cls: Type[_FlatCT], kls: Optional[Type[_CT]]=None) -> Type[_CT]:
+        """Find the corresponding non-flat class.
+
+        The class' bases are searched recursively.
+
+        Parameters
+        ----------
+        kls : :class:`astropy.cosmology.Cosmology` class or None, optional
+            If `None` (default) this class is searched instead of `kls`.
+
+        Raises
+        ------
+        TypeError
+            If more than one non-flat class is found at the same level of the
+            inheritance. This is similar to the error normally raised by
+            python for an inconsistent MRO.
+
+        Returns
+        -------
+        type
+            A Cosmology class that is not a FlatCosmologyMixin subclass.
+        """
+        _kls = cls if kls is None else kls  # Which class to search.
+
+        # Find non-flat classes
+        nonflat: List[Type[_CT]]
+        nonflat = [b for b in _kls.__bases__ if not issubclass(b, FlatCosmologyMixin)]
+
+        if len(nonflat) > 1:
+            raise TypeError(
+                f"cannot create a consistent non-flat class resolution order "
+                f"for {_kls} with bases {nonflat} at the same inheritance level."
+            )
+        elif not nonflat:  # Recurse higher to find non-flat class
+            for base in _kls.__bases__:
+                return cls._get_nonflat_cls(base)
+
+        return nonflat[0]
```

Comment:
If `nonflat` is a `set` not a `list` then indexing wont work:
```suggestion
        return nonflat.pop()
```

Reply:
✅